File: app/src/main/java/net/squanchy/search/SearchItemView.java
package net.squanchy.search;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import net.squanchy.R;
import net.squanchy.imageloader.ImageLoader;
import net.squanchy.search.view.SearchRecyclerView.OnSearchResultClickListener;
import net.squanchy.speaker.domain.view.Speaker;
import net.squanchy.support.lang.Optional;

public class SearchItemView extends LinearLayout {

    private ImageView image;
    private TextView name;

    public SearchItemView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public SearchItemView(Context context, AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public SearchItemView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        super.setOrientation(VERTICAL);
    }

    @Override
    public void setOrientation(int orientation) {
        throw new UnsupportedOperationException("SearchItem doesn't support changing orientation");
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        image = findViewById(R.id.speaker_photo);
        name = findViewById(R.id.speaker_name);
    }

    public void updateWith(Speaker speaker, ImageLoader imageLoader, OnSearchResultClickListener listener) {
        name.setText(speaker.getName());
        updateSpeakerPhotos(speaker, imageLoader);
        setOnClickListener(v -> listener.onSpeakerClicked(speaker));
    }

    private void updateSpeakerPhotos(Speaker speaker, ImageLoader imageLoader) {
        if (imageLoader == null) {
            throw new IllegalStateException("Unable to access the ImageLoader, it hasn't been initialized yet");
        }

        Optional<String> avatarImageURL = speaker.getPhotoUrl();
        if (avatarImageURL.isPresent()) {
            imageLoader.load(avatarImageURL.get()).into(image);
        } else {
            image.setImageResource(R.drawable.ic_no_avatar);
        }
    }
}

File: app/src/main/java/net/squanchy/search/SearchItemView.kt
package net.squanchy.search

import android.content.Context
import android.util.AttributeSet
import android.widget.LinearLayout
import kotlinx.android.synthetic.main.item_search_result_small.view.*
import net.squanchy.R
import net.squanchy.imageloader.ImageLoader
import net.squanchy.search.view.SearchRecyclerView.OnSearchResultClickListener
import net.squanchy.speaker.domain.view.Speaker
import net.squanchy.support.lang.getOrThrow

class SearchItemView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet,
    defStyleAttr: Int = 0,
    defStyleRes: Int = 0
) : LinearLayout(context, attrs, defStyleAttr, defStyleRes) {

    init {
        super.setOrientation(LinearLayout.VERTICAL)
    }

    override fun setOrientation(orientation: Int) {
        throw UnsupportedOperationException("SearchItem doesn't support changing orientation")
    }

    fun updateWith(speaker: Speaker, imageLoader: ImageLoader, listener: OnSearchResultClickListener) {
        speakerName.text = speaker.name
        updateSpeakerPhotos(speaker, imageLoader)
        setOnClickListener { listener.onSpeakerClicked(speaker) }
    }

    private fun updateSpeakerPhotos(speaker: Speaker, imageLoader: ImageLoader?) {
        if (imageLoader == null) {
            throw IllegalStateException("Unable to access the ImageLoader, it hasn't been initialized yet")
        }

        val avatarImageURL = speaker.photoUrl
        if (avatarImageURL.isDefined()) {
            imageLoader.load(avatarImageURL.getOrThrow())
                .placeholder(R.drawable.ic_avatar_placeholder)
                .error(R.drawable.ic_no_avatar)
                .into(speakerPhoto)
        } else {
            speakerPhoto.setImageResource(R.drawable.ic_no_avatar)
        }
    }
}
